dotnet/src/Microsoft.Agents.AI.AGUI/AGUIChatClient.cs

```diff
@@ -0,0 +1,266 @@
+        {
+            updates.Add(update);
+        }
+
+        return updates.ToChatResponse();
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<ChatResponseUpdate> GetStreamingResponseAsync(
+        IEnumerable<ChatMessage> messages,
+        ChatOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        await foreach (var update in this._innerClient.GetStreamingResponseAsync(messages, options, cancellationToken).ConfigureAwait(false))
+        {
+            // We bypass the underlying FunctionInvokingChatClient for any server executed function.
+            // This is because it will try to invoke it and fail because it can't find the function.
+            // There are ways in which we could make it work and return the existing result (with a FunctionInvoker) but we would need to very
+            // carefully manage the buffering to preserve the order of updates.
+            // This is because FunctionInvokingChatClient will process all updates in a turn and then trigger the function invocation process,
+            // meaning that any text update after a server function call would jump out of order on the stream.
+            // Instead, we buffer any server function updates internally, and after each update we check if we have something buffered to yield.
+            // This way we preserve the order and don't have to manage it inside function invoking client.
+            if (this._executedServerFunctionStream.Count > 0)
+            {
+                foreach (var executedFunctionUpdate in this._executedServerFunctionStream)
+                {
+                    yield return executedFunctionUpdate;
+                }
+                this._executedServerFunctionStream.Clear();
+            }
+
+            yield return update;
+        }
+
+        // In case the functions happened at the end of the stream, yield them now
+        foreach (var executedFunctionUpdate in this._executedServerFunctionStream)
+        {
+            yield return executedFunctionUpdate;
+        }
+        this._executedServerFunctionStream.Clear();
+    }
+
+    /// <inheritdoc />
+    public object? GetService(Type serviceType, object? serviceKey = null)
+    {
+        if (typeof(FunctionInvokingChatClient) == serviceType)
+        {
+            return (FunctionInvokingChatClient)this._innerClient.GetService(serviceType, serviceKey)!;
+        }
+        return this._innerClient.GetService(serviceType, serviceKey);
+    }
+
+    private sealed class AGUIChatClientHandler : IChatClient
+    {
+        private readonly AGUIHttpService _httpService;
+        private readonly JsonSerializerOptions _jsonSerializerOptions;
+
+        private readonly List<ChatResponseUpdate> _serverFunctionUpdates = [];
+        private readonly ILogger _logger;
+
+        private List<ChatMessage>? _currentMessages;
```

Comment:
And here as well.

Is it guaranteed that one of these chat client instances is only ever used for a single request at a time?